### ISSUE TITLE
A bit slow when adding refs from a moderate (2k) bib file

### ISSUE THREAD

**Issue by Robinlovelace** [crsh/citr#23]
As shown below, it's not when firing-up the reference to choose from, but after 'Insert citation' has been clicked. Hoping this will help assess ideas to speed-it-up because **citr** is a really fantastic add-on that provides the missing piece in the jigsaw to provide the ultimate reproducible research environment.

**Comment by Robinlovelace** [comment]
![peek 2017-12-13 23-38](https://user-images.githubusercontent.com/1825120/33968282-210e46b6-e05f-11e7-82ae-679bf7ea37b1.gif)

**Comment by Robinlovelace** [comment]
P.s. that was with 

```
packageVersion("citr")
[1] ‘0.2.0.9047’
```

**Comment by Robinlovelace** [comment]
Just tested on latest version and having the same issue. Still using it though as it's a great package. Should I use a smaller .bib file?

**Comment by crsh** [comment]
Hi Robin, thanks for the report. That's definitely an undesirably slow response. Could you send me the bib-file to play with (either via e-mail or by uploading it here)? I'll try exploring what's causing the delay and how to make the code more efficient.

**Comment by Robinlovelace** [comment]
Here you go (removed now).

**Comment by Robinlovelace** [comment]
Also: I'm running Zotero and I think that may be part of it - it's not just a bib file it's interactive with as far as I can tell (loads of files like `LDJHPOAPOIJJDOIJVLA.bib` appear ephemerally in the repo when I run **citr**).

**Comment by crsh** [comment]
Yes, these temporary files are currently necessary because the functions I rely on to read the `.bib`-files fail when the files get too large. Hence, I need to split them before reading and merging them. This may be part of the issue. However, `insert_citation()` should only read your Zotero library once (and cache it) unless you click "Reload database". I suspect the slow response is related to the `.bib`-file that is updated if necessary every time a new citation is inserted from Zotero.

P.S.: I've downloaded the file, you can delete it, if you want.

**Comment by Robinlovelace** [comment]
OK that makes sense. Can you reproduce this lethargic snail-like speed of citation delivery?

**Comment by Robinlovelace** [comment]
Just an unrelated update I thought you might like: just discovered the `tidy_bib_file()` function and am absolutely stoked at how well it works. I've been looking for this functionality for literally years. Awesome work there!

```
> citr::tidy_bib_file(rmd_file = "README.Rmd", messy_bibliography = "~/allrefs.bib", "who.bib")
Removing 1999 unneeded bibliography entries.
Writing 6 Bibtex entries ... OK
Results written to file 'who.bib'
Warning message:
In do_read_bib(file, encoding = .Encoding, srcfile) :
  encoding 'native.enc' will be ignored
```

**Comment by crsh** [comment]
Sorry, I haven't found the time to take a closer look at your original issue but I'm glad to hear you like that function! I think I will add this as another addin for the next release.

**Comment by crsh** [comment]
Hi Robin, as I suspected, this was due to the fact that the addin updates the bibliography file when inserting citations form Zotero. This can take a little while when the file is large. I have now added a progress indicator to call attention to this fact. Also, the contents of the file should now be cached so the addin should only have to read the file once. The second time you insert a citation should be much faster. Let me know if it works.
